lib/tracing-limit/src/lib.rs

```diff
@@ -10,165 +10,200 @@ use std::{
+                    values.record(&mut fields);
+                    extensions.insert(fields);
+                }
+            };
         }
+        self.inner.on_record(id, values, ctx);
+    }
 
-        true
+    #[inline]
+    fn on_follows_from(&self, span: &span::Id, follows: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_follows_from(span, follows, ctx);
     }
 
-    fn on_event(&self, event: &Event, _ctx: Context<S>) {
-        if is_limited(event.metadata()) {
-            let mut limit_visitor = LimitVisitor::default();
-            event.record(&mut limit_visitor);
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let metadata = event.metadata();
 
-            if let Some(limit) = limit_visitor.limit {
-                let start = Instant::now();
+        // if the event is not rate limited, just pass through
+        if !is_limited(metadata) {
+            return self.inner.on_event(event, ctx);
+        }
 
-                let id = event.metadata().callsite();
+        let mut limit_visitor = LimitVisitor::default();
+        event.record(&mut limit_visitor);
+
+        let component_name = {
+            let mut scope = ctx
+                .lookup_current()
+                .into_iter()
+                .flat_map(|span| span.from_root().chain(std::iter::once(span)));
+
+            scope.find_map(|span| {
+                let extensions = span.extensions();
+                extensions
+                    .get::<RateLimitedSpanKeys>()
+                    .and_then(|fields| fields.component_name.clone())
+            })
+        };
+
+        let id = RateKeyIdentifier {
+            callsite: metadata.callsite(),
+            rate_limit_key: limit_visitor.key.clone(),
+            component_name,
+        };
+
+        let events = self.events.read().expect("lock poisoned!");
+
+        // check if the event exists within the map, if it does
+        // that means we are currently rate limiting it.
+        if let Some(state) = events.get(&id) {
+            let start = state.start.unwrap_or_else(Instant::now);
+
+            // check if we are still rate limiting
+            if start.elapsed().as_secs() < state.limit {
+                let prev = state.count.fetch_add(1, Ordering::Relaxed);
+                if prev == 1 {
+                    // output first rate limited log
+                    let message = match limit_visitor.key {
+                        None => {
+                            format!("Internal log [{}] is being rate limited.", state.message)
+                        }
+                        Some(key) => format!(
+                            "Internal log [{} internal_log_rate_key={}].",
+                            state.message, key,
+                        ),
+                    };
+
+                    self.create_event(&ctx, metadata, message, state.limit);
+                }
+            } else {
+                // done rate limiting
```

Comment:
I think we have to call `inner.on_event` here, for example, if `log_rate_secs` is 1 and another event emitted 2 seconds later, it would be rate-limited.

Reply:
Good catch, I'll look into that. I also realized there could be a race condition here too if two log events come in at "same time" where they both check for state, find none, output, and insert into the state map.